### PR TITLE
Custom URL support for analytics-dashboard

### PR DESCRIPTION
**Pull request by fazlan-nazeem**

### CODE REVIEW

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimAdminApiClient.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.wso2.analytics.apim.idp.client;
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimAdminApiClient.java]: license?

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/dto/CustomUrlInfoDevPortalDTO.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package org.wso2.analytics.apim.idp.client.dto;
```
Comment (tharikaGitHub) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/dto/CustomUrlInfoDevPortalDTO.java]: Shall we add the license header here?

Comment (TanyaM) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/dto/CustomUrlInfoDevPortalDTO.java]: Is this Auto gen? If not shall we add the license?

Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/dto/CustomUrlInfoDevPortalDTO.java]: It has been added in the last commit.

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 3]
```diff
@@ -623,10 +705,20 @@ private void registerApplication(String appContext, String clientName, String km
...
+                if (isCustomUrlApplicable) {
+                    appContext = appContext + "_" + customUrlInfo.getTenantDomain();
+                    tenantDomain = customUrlInfo.getTenantDomain();
+                }
+                this.oAuthAppInfoMap.put(appContext, oAuthApplicationInfo);
+                this.customUrlInfoMap.put(customUrlInfo.getTenantDomain(), customUrlInfo);
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("OAuth2 application created: " + oAuthApplicationInfo.toString());
                 }
+                oAuthAppDAO.insertSystemApp(dcrClientInfoResponse, clientName,  tenantDomain);
```
Comment (RAVEENSR) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]: Why we need to insert the app to system apps table here? When we make the DCR call, APIM inserts this to the system apps table right?

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 4]
```diff
@@ -623,10 +705,20 @@ private void registerApplication(String appContext, String clientName, String km
                 OAuthApplicationInfo oAuthApplicationInfo = new OAuthApplicationInfo(
                         clientName, dcrClientInfoResponse.getClientId(), dcrClientInfoResponse.getClientSecret()
                 );
-                this.oAuthAppInfoMap.replace(appContext, oAuthApplicationInfo);
+                String tenantDomain = "carbon.super";
```
Comment (RAVEENSR) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]:
```suggestion
                String tenantDomain = SUPER_TENANT_DOMAIN;
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 5]
```diff
@@ -266,35 +298,44 @@ private String extractTenantDomainFromUserName(String username) throws IdPClient
 
     @Override
     public Map<String, String> login(Map<String, String> properties) throws IdPClientException {
+
         Map<String, String> returnProperties = new HashMap<>();
+        CustomUrlInfo customUrlInfo = getCustomUrlInfo(properties.getOrDefault(IdPClientConstants.DOMAIN,
+                "carbon.super"));
```
Comment (RAVEENSR) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]:
```suggestion
                SUPER_TENANT_DOMAIN));
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 6]
```diff
@@ -423,15 +476,35 @@ private String extractTenantDomainFromUserName(String username) throws IdPClient
...
+
         Map<String, String> returnProperties = new HashMap<>();
         String oAuthAppContext = appContext.split("/\\|?")[0];
         if (!this.oAuthAppInfoMap.containsKey(oAuthAppContext)) {
             oAuthAppContext = ApimIdPClientConstants.DEFAULT_SP_APP_CONTEXT;
         }
-        OAuthApplicationInfo oAuthApplicationInfo = this.oAuthAppInfoMap.get(oAuthAppContext);
+        OAuthApplicationInfo oAuthApplicationInfo;
+        String baseUrl;
+        String tenantDomain = properties.getOrDefault(IdPClientConstants.DOMAIN, "carbon.super");
```
Comment (RAVEENSR) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]:
```suggestion
        String tenantDomain = properties.getOrDefault(IdPClientConstants.DOMAIN, SUPER_TENANT_DOMAIN);
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java (hunk 3 above)
Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]: No, it doesn't. So we have to write to this table from here.

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java (hunk 4 above)
Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]: done

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java (hunk 5 above)
Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]: done

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java (hunk 6 above)
Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java]: done

**Pull request merged**